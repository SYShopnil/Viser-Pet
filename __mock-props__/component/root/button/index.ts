import {
  BtnColorSchema,
  BtnSize,
  INormalButton,
  IRedirectButton,
} from '@src/types/root';

export const mock_data_normalButton: INormalButton[] = [
  {
    colorSchema: BtnColorSchema.BrownBgTextWhite,
    btnSize: BtnSize.Small,
    btnText: 'Normal Button One',
    hoverColorSchema: BtnColorSchema.WhiteBgTextBrown,
    isUpperCased: false,
    clickHandler: (e) => {
      console.log(`Clicked Normal button`);
    },
    dataTestId: 'btn-normal-one',
  },
  {
    colorSchema: BtnColorSchema.WhiteBgTextBrown,
    btnSize: BtnSize.Small,
    btnText: 'Normal Button Two',
    hoverColorSchema: BtnColorSchema.BrownBgTextWhite,
    isUpperCased: false,
    clickHandler: (e) => {
      console.log(`Clicked Normal button`);
    },
    dataTestId: 'btn-normal-two',
  },
  {
    colorSchema: BtnColorSchema.WhiteBgTextGray,
    btnSize: BtnSize.Small,
    btnText: 'Normal Button Three',
    hoverColorSchema: BtnColorSchema.WhiteBgTextBrown,
    isUpperCased: false,
    clickHandler: (e) => {
      console.log(`Clicked Normal button`);
    },
    dataTestId: 'btn-normal-three',
  },
];

export const mock_data_redirectButton: IRedirectButton[] = [
  {
    btnSize: BtnSize.Small,
    btnText: 'Redirect Button one',
    colorSchema: BtnColorSchema.BrownBgTextWhite,
    hoverColorSchema: BtnColorSchema.WhiteBgTextBrown,
    isOpenNewTab: true,
    redirectLink: 'https://chat.openai.com/',
    isUpperCased: true,
    dataTestId: 'Redirect Button One',
  },
  {
    btnSize: BtnSize.Large,
    btnText: 'Redirect Button two',
    colorSchema: BtnColorSchema.WhiteBgTextBrown,
    hoverColorSchema: BtnColorSchema.BrownBgTextWhite,
    isOpenNewTab: true,
    redirectLink: 'https://chat.openai.com/',
    isUpperCased: false,
    dataTestId: 'Redirect Button Two',
  },
  {
    btnSize: BtnSize.Medium,
    btnText: 'Redirect Button Three',
    colorSchema: BtnColorSchema.WhiteBgTextGray,
    hoverColorSchema: BtnColorSchema.WhiteBgTextGray,
    isOpenNewTab: true,
    redirectLink: 'https://chat.openai.com/',
    isUpperCased: false,
    dataTestId: 'Redirect Button Three',
  },
  {
    btnSize: BtnSize.Small,
    btnText: 'Redirect Button Four',
    colorSchema: BtnColorSchema.BrownBgTextWhite,
    hoverColorSchema: BtnColorSchema.WhiteBgTextBrown,
    isOpenNewTab: true,
    redirectLink: 'https://chat.openai.com/',
    isUpperCased: false,
    dataTestId: 'Redirect Button Five',
  },
];
