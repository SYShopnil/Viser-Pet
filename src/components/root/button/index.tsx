import {
  BtnColorSchema,
  BtnSize,
  IButton,
  INormalButton,
  IRedirectButton,
} from '@src/types/root';
import Link from 'next/link';
import React, { MouseEvent } from 'react';
import ModuleStyle from './button.module.scss';

//Common layout
const BaseButton = ({
  btnSize,
  btnText,
  colorSchema,
  hoverColorSchema,
  isUpperCased,
}: IButton) => {
  /**Not fixed */
  const common_btn_txt_wrapper_style = `inline-block !font-secondary`;
  const common_btn_main_wrapper_style = `w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]`;
  const bg_theme: string =
    Object.keys(BtnColorSchema).find((color: string) => color == colorSchema) ||
    'BrownBgTextWhite';

  const bg_hover_theme: string =
    Object.keys(BtnColorSchema).find(
      (color: string) => color == hoverColorSchema
    ) || 'WhiteBgTextBrown';

  const btn_size: string =
    Object.values(BtnSize).find((size: string) => btnSize == size) ||
    BtnSize.Small;

  const isUpperCaseOrCapitalized = isUpperCased
    ? 'text-transform-uppercase'
    : 'text-transform-capitalized';

  return (
    // button main wrapper
    <span
      className={`${common_btn_main_wrapper_style}   ${ModuleStyle[bg_theme]} ${
        ModuleStyle['mainWrapper']
      } ${ModuleStyle[`${bg_hover_theme}Hover`]} ${
        ModuleStyle[`${btn_size}-wrapper`]
      }`}
    >
      {/* btn text part wrapper */}
      <span
        className={`${common_btn_txt_wrapper_style} ${ModuleStyle[btn_size]} ${ModuleStyle[isUpperCaseOrCapitalized]}`}
      >
        {btnText}
      </span>
    </span>
  );
};

//Normal Button
export const Button = (props: INormalButton) => {
  return (
    <span
      onClick={(
        e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
      ) => props.clickHandler?.(e)}
      data-testid={props.dataTestId}
    >
      <BaseButton {...props} />
    </span>
  );
};

//Redirect Button
Button.RedirectButton = (props: IRedirectButton) => {
  return (
    <Link href={props.redirectLink} passHref>
      <a
        target={props.isOpenNewTab ? '_blank' : '_self'}
        data-testid={props.dataTestId}
      >
        <BaseButton {...props} />
      </a>
    </Link>
  );
};
export default Button;
