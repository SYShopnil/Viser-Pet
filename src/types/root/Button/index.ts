export enum BtnColorSchema {
  BrownBgTextWhite = 'BrownBgTextWhite',
  WhiteBgText = 'WhiteBgText',
}
export enum BtnSize {
  Small = 'btn-sm',
  Medium = 'btn-md',
  Large = 'btn-lg',
  ExtraLarge = 'btn-ex-lg',
}
export interface IButton {
  btnText: string;
  colorSchema: BtnColorSchema;
  btnSize: BtnSize;
  hoverColorSchema: BtnColorSchema;
}
export interface INormalButton extends IButton {
  clickHandler?: (e: MouseEvent, ...a: any[]) => void;
}
export interface IRedirectButton extends IButton {
  isOpenNewTab: boolean;
  redirectLink: string;
}
