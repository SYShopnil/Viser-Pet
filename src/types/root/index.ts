export type { IButton, INormalButton, IRedirectButton } from './Button/';

export { BtnSize, BtnColorSchema } from './Button/';
