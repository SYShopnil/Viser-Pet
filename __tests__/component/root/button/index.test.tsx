import {
  mock_data_normalButton,
  mock_data_redirectButton,
} from '@root/__mock-props__/component/root/button';
import { Button } from '@src/components/root';
import { describe } from 'node:test';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('When Normal button has rendered', () => {
  it.each(mock_data_normalButton)(
    `Expect successfully rendered`,
    (btnMockData) => {
      const { getByTestId } = render(<Button {...btnMockData} />);
      const btn = getByTestId(btnMockData.dataTestId || '');
      expect(btn).toBeInTheDocument();
      expect(btn).toMatchSnapshot();
    }
  );
});

describe('When Redirect Button has rendered', () => {
  it.each(mock_data_redirectButton)(
    'Expect successfully rendered',
    (mockButtonData) => {
      const { getAllByTestId } = render(
        <Button.RedirectButton {...mockButtonData} />
      );
      const redirectButton = getAllByTestId(mockButtonData.dataTestId || '');
    }
  );
});
// To Do
/**
 * 1. Check button size
 * 2. Check button title works perfectly or not
 * 3. Check isUppersCase worked or not
 *  4. Check bg theme class
 *
 */
