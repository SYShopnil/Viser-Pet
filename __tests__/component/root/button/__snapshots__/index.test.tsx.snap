// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect  successfully rendered 1`] = `
<span
  data-testid="btn-normal-one"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   BrownBgTextWhite mainWrapper WhiteBgTextBrownHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button One
    </span>
  </span>
</span>
`;

exports[`Expect  successfully rendered 2`] = `
<span
  data-testid="btn-normal-two"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   WhiteBgTextBrown mainWrapper BrownBgTextWhiteHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button Two
    </span>
  </span>
</span>
`;

exports[`Expect  successfully rendered 3`] = `
<span
  data-testid="btn-normal-three"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   WhiteBgTextGray mainWrapper WhiteBgTextBrownHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button Three
    </span>
  </span>
</span>
`;

exports[`Expect successfully rendered 1`] = `
<span
  data-testid="btn-normal-one"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   BrownBgTextWhite mainWrapper WhiteBgTextBrownHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button One
    </span>
  </span>
</span>
`;

exports[`Expect successfully rendered 2`] = `
<span
  data-testid="btn-normal-two"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   WhiteBgTextBrown mainWrapper BrownBgTextWhiteHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button Two
    </span>
  </span>
</span>
`;

exports[`Expect successfully rendered 3`] = `
<span
  data-testid="btn-normal-three"
>
  <span
    class="w-[100%] h-[auto]  flex justify-center items-center cursor-pointer  rounded-[1.875rem]   WhiteBgTextGray mainWrapper WhiteBgTextBrownHover btn-sm-wrapper"
  >
    <span
      class="inline-block !font-secondary btn-sm text-transform-capitalized"
    >
      Normal Button Three
    </span>
  </span>
</span>
`;
